### PR TITLE
Separate codes one element forms

### PR DESCRIPTION
**Pull request by iafilatov**
This adds separate C82X codes for cases when a trailing comma is required in an enclosure with a single element. The reason is that sometimes people prefer not to have to add comas when for the enclosure is used mostly for grouping. For example:
```python
logger.info(
    'This is a long and excessively detailed '
    'possibly multi-line '
    'log message.'  # C822 missing trailing comma in one element enclosure
)
```
C822 can be exluded and calls with more than one arg get a separate C812 error code and can be left on:
```python
logger.info(
    'log message',
    extra={'extra': 'data'}  # C812 missing trailing comma in enclosure
)
```

This also adds separate codes for function definitions and single element collections.

### CODE REVIEW

**Comment by graingert** [comment]
Not sure about this one, as that's working as designed.

Can you rebase?

**Comment by graingert** [comment]
@iafilatov sorry about the delay on this, as for some reason I was not watching this repo!

**Comment by graingert** [comment]
when you add a second arg to:
```python
foo(
    'ham'
    'spam'
)
```                      
you get
```python
foo(
    'ham'
    'spam',
    'eggs',
)
```

and the associated diff spam:
```patch
--- before.py
+++ after.py
@@ -1,4 +1,5 @@
 foo(
     'ham'
-    'spam'
+    'spam',
+    'eggs',
 )
```

However adding more text to:

```python
foo(
    'ham'
    'spam',
)
```        

gets you:

```python
foo(
    'ham'
    'spam'
    'eggs',
)
```

and just as much diff spam:
```patch
--- before.py
+++ after.py
@@ -1,4 +1,5 @@
 foo(
     'ham'
-    'spam',
+    'spam'
+    'eggs',
 )
```

**Comment by gordon-rehabstudio** [comment]
mixing automatic string concatenation and multiple arguments within one set of () is of the devil

File: test/test_checks.py [hunk 1]
```diff
@@ -4,10 +4,22 @@
...
+C817 = 'C817 missing trailing comma in function def'
+C818 = 'C818 missing trailing comma in collection'
+
+C822 = 'C822 missing trailing comma in one element enclosure'
+C823 = 'C823 missing trailing comma in Python 3 in one element enclosure'
+C824 = 'C824 missing trailing comma in Python 2 in one element enclosure'
+C825 = 'C825 missing trailing comma in Python 3.5+ in one element enclosure'
+C826 = 'C826 missing trailing comma in Python 3.6+ in one element enclosure'
+C827 = 'C827 missing trailing comma in funtion def with one argument'
+C828 = 'C828 missing trailing comma in one element collection'
```
Comment (graingert) [inline test/test_checks.py]: I only lay claim to C81X, You'll have to change this to C8XX

https://github.com/flake8-commas/flake8-commas/blob/master/setup.py#L38

Comment (adamchainz) [inline test/test_checks.py]: "I" = "flake8-commas"

**Comment by graingert** [comment]
Is there a way to make it so it's only special cased if the single element is a line broken concatenated element?

**Comment by iafilatov** [comment]
Hm, not sure I follow. Could you give an example of what should/should not be special cased?

**Comment by graingert** [comment]
Eg:

```python
foo(
    'hamspam'  # C812
)
```


```python
foo(
    'ham'
    'spam'  # C822
)
```

**Comment by iafilatov** [comment]
Technically both are function calls with one argument.  so I think both should be other than C812. That's the purpose of this PR. Maybe the second one could get a different C82X… not sure why, though. Do you have a use case?

**Comment by graingert** [comment]
Because in the first example a trailing comma is definitely needed. I thought this PR was about the case were people use parenth_forms to concat strings

**Comment by iafilatov** [comment]
I see. Maybe I didn't explain myself well. The original intention was to give all one element enclosures separate codes since sometimes it's preferable to allow them (despite diff spam).  String concatenation is one case.

**Comment by iafilatov** [comment]
So what about this? I don't see any problems because it doesn't relax any rules, just gives different codes to a subset of them.

**Comment by graingert** [comment]
Sure but it's designed so people can ignore C822 if they want.

For now I want to only special case single element brackets and implicit
concatenation of strings.

**Comment by mst-ableton** [comment]
Looking forward to seeing the special casing go through. Any immediate plans for that? I can attempt a PR but it would take me a bit to get up to speed.

**Comment by sanjioh** [comment]
strong +1 to this, as it’s the single thing I feel this plugin is missing.
Consider this use case as well:

```
’,’.join(
    [item for item in very_long_expression]
)
```

**Comment by extsoft** [comment]
@iafilatov @graingert @adamchainz I reported #55 and looks like this PR solves it. Can I make some efforts to speed up the process of merging this PR?

**Comment by graingert** [comment]
Yeah, it needs rebasing and changing so it only applies if the single element is an implicitly concatenated string

**Comment by extsoft** [comment]
@graingert I see two options then:
1. wait for the author
2. create another PR based on this one

What is preferred?

**Comment by extsoft** [comment]
As I see, #55 is no solved by this PR. So, I'll create a separate one.

**Comment by extsoft** [comment]
@iafilatov if you create an issue to describe the changes of this PR, I can try to get the logic merged.

**Comment by graingert** [comment]
Closing due to inactivity

**Pull request closed without merging**